src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs

```diff
@@ -506,19 +506,24 @@ private Task ProcessUnknownFrameAsync()
                 case RequestHeaderParsingState.Trailers:
                     // trailers
                     // TODO figure out if there is anything else to do here.
-                    return Task.CompletedTask;
+                    return;
                 default:
                     Debug.Fail("Unexpected header parsing state.");
                     break;
             }
 
             InputRemaining = HttpRequestHeaders.ContentLength;
 
+            // If the stream is complete after receiving the headers then run OnEndStreamReceived.
+            // If there is a bad content length then this will throw before the request delegate is called.
+            if (isCompleted)
+            {
+                await OnEndStreamReceived();
+            }
```

Comment:
This matches HTTP/2 behavior which validates content-length if end stream is received with HEADERS flag. This prevents the RequestDelegate from executing.